panningDirection parameter added to InteractiveViewer

NazarenoCavazzon opened the pull request:
It adds a new Axis? parameter to the Interactive Viewer called panningDirection that let's you lock the panning to a certain axis.

Fixes https://github.com/flutter/flutter/issues/108935

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

NazarenoCavazzon committed:
panning direction working

NazarenoCavazzon committed:
Tests added

NazarenoCavazzon committed:
Fix

justinmc commented:
@NazarenoCavazzon Thanks for the PR!  Overall I think this parameter makes sense, I agree that alignPanAxis doesn't quite do what you want.

It looks like you ran a formatter on these files, though.  Would you be able to undo the formatter changes so that this is easier to review?

NazarenoCavazzon commented:
Yep, didn't saw it hahahaha, I'll undo it

NazarenoCavazzon commented:
I have also a question, do I need to update the flutter version or make a document like [this one](https://docs.google.com/document/d/1bcPV8yNAETmlxTFTMpO_JcJ7-yMoF8qaCPa884hFmXA/edit?usp=drivesdk&resourcekey=0-73tg8sZOVf0YXBmz-MLwQw) explaining the changes and the tests I added?

NazarenoCavazzon committed:
Revert formatting

NazarenoCavazzon commented:
@justinmc I reverted the formatting, let me know if you need something

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -366,6 +366,123 @@ void main() {
+      await gesture.up();
+      await tester.pumpAndSettle();
+
+      // Translation has only happened along the y axis (the default axis when
```
Should this say "x axis" or am I misunderstanding?

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -366,6 +366,123 @@ void main() {
+      expect(translation.y, 0.0);
+    });
+
+     testWidgets('panningDirection do not allow panning in vertical direction on vertical gesture', (WidgetTester tester) async {
```
"do not" => "does not"

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -366,6 +366,123 @@ void main() {
+      await gesture.up();
+      await tester.pumpAndSettle();
+
+      // Translation happened didn't happened, because the only axis allowed to
```
"happened didn't happened" => "didn't happen"

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -366,6 +366,123 @@ void main() {
+      await tester.pumpAndSettle();
+
+      // Translation happened didn't happened, because the only axis allowed to
+      //do panning is the horizontal.
```
Space after the `//`.

justinmc reviewed (commented):
Thanks for undoing the formatting, much easier to read now!  A few small comments below.

If someone specified panningDirection, then there is no reason to specify alignPanAxis too, right?  Maybe add an `assert` to make sure that both aren't set at the same time?

Or, maybe the two parameters should be combined?  What if there was just one parameter, an enum (maybe called panAxis?) with values horizontal, vertical, both, and aligned.

To answer your question about writing a design doc, I think it's probably not necessary since this feature is relatively small and not super controversial.  It's always welcome if you want to write one though :)

NazarenoCavazzon commented:
I like more the enum approach, it makes more sense

NazarenoCavazzon committed:
Made alignPanAxis an Enum to integrate panningDirection and alignPanAxis into one

NazarenoCavazzon commented:
@justinmc I made the alignPanAxis and Enum.

NazarenoCavazzon committed:
Delete trailing spaces and replacing alignPanAxis: true

NazarenoCavazzon commented:
Everything ready 2 review

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -152,18 +152,17 @@ class InteractiveViewer extends StatefulWidget {
-  /// If true, panning is only allowed in the direction of the horizontal axis
-  /// or the vertical axis.
+  /// When set to [PanAxis.aligned] panning is only allowed in the horizontal
+  /// axis or the vertical axis, diagonal panning not allowed.
```
"diagonal panning not allowed" => "diagonal panning is not allowed"

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -152,18 +152,17 @@ class InteractiveViewer extends StatefulWidget {
+  /// When set to [PanAxis.vertical] or [PanAxis.horizontal] panning is only
+  /// allowed in the specified axis. For example, if set to [PanAxis.vertical]
+  /// if you try to pan in the horizontal axis the InteractiveViewer will not
+  /// pan. And if you try to pan in diagonal directions the InteractiveViewer will
```
Can you change this to not use the word "you"?  See https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-the-passive-voice-recommend-do-not-require.

Otherwise the general idea makes sense to me :+1:

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -1290,3 +1306,20 @@ Axis? _getPanAxis(Offset point1, Offset point2) {
+/// This enum is used to specify the behavior of the [InteractiveViewer] when
+/// the user drags the viewport.
+enum PanAxis{
+  /// The user can pan the viewport along the horizontal axis.
```
"can" => "can only"

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -1290,3 +1306,20 @@ Axis? _getPanAxis(Offset point1, Offset point2) {
+  /// The user can pan the viewport along the horizontal axis.
+  horizontal,
+
+  /// The user can pan the viewport along the vertical axis.
```
can" => "can only"

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -1290,3 +1306,20 @@ Axis? _getPanAxis(Offset point1, Offset point2) {
+  aligned,
+
+  /// The user can pan the viewport along both the horizontal and vertical axes.
+  both,
```
Nit: I feel like there's a better word for this instead of "both"... Maybe "free"?

NazarenoCavazzon replied:
Perfect, both sounds kind of wacky, let's use free!

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -61,7 +61,7 @@ class InteractiveViewer extends StatefulWidget {
     super.key,
     this.clipBehavior = Clip.hardEdge,
-    this.alignPanAxis = false,
+    this.alignPanAxis = PanAxis.both,
```
I wonder if it might be better to create a completely new parameter ("panAxis"?) so that we can use Flutter's normal deprecation flow.  That's probably the smoothest approach for anyone that might be using alignPanAxis right now.

Plus, I think panAxis is a better name than alignPanAxis now.  What do you think?

Here's an example of another deprecated parameter in another widget:

https://github.com/flutter/flutter/blob/6e95067273da1f886aa64cf475015f999f6ce817/packages/flutter/lib/src/material/app_bar.dart#L194-L198

NazarenoCavazzon replied:
I thought about it but I waited until your review cause I wanted to hear the name of the new parameter by you. 

I think that's the best way to do it.

NazarenoCavazzon replied:
@justinmc quick question, how do I know the version in which this parameter is being deprecated?

justinmc replied:
Check the latest tag here: https://github.com/flutter/flutter/tags

NazarenoCavazzon replied:
Found out today, now I'm using the correct tag 👌🏻

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -295,7 +295,7 @@ void main() {
             body: Center(
               child: InteractiveViewer(
-                alignPanAxis: true,
+                alignPanAxis: PanAxis.aligned,
```
Can you update the name of this test and the one below?  It says alignPanAxis but should probably say "PanAxis.aligned" now.

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -366,6 +366,123 @@ void main() {
       expect(translation.y, 0.0);
     });
 
+    testWidgets('panningDirection allows panning in one direction only for diagonal gesture', (WidgetTester tester) async {
```
I think you forgot to update this: "panningDirection" => "PanAxis.horizontal"

Also instead of saying "one direction only" maybe say "the horizontal direction only".

justinmc replied:
Same for the other tests below.

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -519,7 +636,7 @@ void main() {
             body: Center(
               child: InteractiveViewer(
-                alignPanAxis: true,
+                alignPanAxis: PanAxis.aligned,
```
I assume this test is also true for PanAxis.horizontal and vertical.  Probably no need to test those too though...

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -366,6 +366,123 @@ void main() {
+          home: Scaffold(
+            body: Center(
+              child: InteractiveViewer(
+                alignPanAxis: PanAxis.horizontal,
```
Should you also test PanAxis.vertical?  I know it's the same thing, but maybe just duplicate one of these tests and modify it for `vertical`?

justinmc reviewed (commented):
I really like the enum approach, thanks for doing that!

The main thing is just my comment about using a new name and deprecating alignPanAxis.

NazarenoCavazzon committed:
PR's review changes

- Modified the new parameter to be called panAxis
- Restored alignPanAxis and added the deprecated annotation
- Default value of panAxis changed from both to free
- Modified some comments
- Added 3 more test for panAxis.vertical and 1 for panAxis.free

NazarenoCavazzon committed:
Merge remote-tracking branch 'upstream/master' into feat/interactive-viewer-panning-direction

NazarenoCavazzon committed:
Changed deprecated message

NazarenoCavazzon committed:
Update interactive_viewer.dart

NazarenoCavazzon committed:
Update interactive_viewer.dart

NazarenoCavazzon committed:
Correct version xd

NazarenoCavazzon committed:
Updated alignPanAxis to panAxis in the example

NazarenoCavazzon commented:
Ready 2 review

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -82,7 +87,7 @@ class InteractiveViewer extends StatefulWidget {
     this.scaleFactor = 200.0,
     this.transformationController,
     required Widget this.child,
-  }) : assert(alignPanAxis != null),
```
Can you keep this assert here for now?  Once alignPanAxis is removed then this assert should be removed.

NazarenoCavazzon replied:
Added it again

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -169,8 +181,25 @@ class InteractiveViewer extends StatefulWidget {
+  )
   final bool alignPanAxis;
 
+  /// When set to [PanAxis.aligned] panning is only allowed in the horizontal
```
Add a comma: "When set to [PanAxis.aligned],"

NazarenoCavazzon replied:
Added

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -566,9 +595,26 @@ class _InteractiveViewerState extends State<InteractiveViewer> with TickerProvid
+          alignedTranslation = _alignAxis(translation, Axis.vertical);
+          break;
+        case PanAxis.aligned:
+          alignedTranslation = _alignAxis(translation, _panAxis!);
```
Can you update the comment at the definition of _panAxis?  It still references alignPanAxis.

Also it's kind of confusing that there is both `panAxis` and `_panAxis`, but they mean different things and have different types...  Maybe if we just change _panAxis to _panningAxis or _currentAxis or something?  Up to you.

NazarenoCavazzon replied:
All done!

justinmc reviewed (approved):
LGTM with nits :+1: 

Can you search through the code and make sure alignPanAxis isn't referenced anywhere in comments or anything?  I think I found one case below.

pdblasi-google reviewed (approved):
LGTM

NazarenoCavazzon committed:
PR Fixes

NazarenoCavazzon committed:
Merge remote-tracking branch 'upstream/master' into feat/interactive-viewer-panning-direction

NazarenoCavazzon committed:
Merge remote-tracking branch 'upstream/master' into feat/interactive-viewer-panning-direction

NazarenoCavazzon committed:
Merge remote-tracking branch 'upstream/master' into feat/interactive-viewer-panning-direction

justinmc commented:
Thanks for your persistence on this PR and for fixing those last few comments!  This will be merged when the build is green.

NazarenoCavazzon commented:
Thanks for reviewing it!
